experimental/metal/staging_buffer.h

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_EXPERIMENTAL_METAL_STAGING_BUFFER_H_
+#define IREE_EXPERIMENTAL_METAL_STAGING_BUFFER_H_
+
+#import <Metal/Metal.h>
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// Size, in bytes, of the shared storage mode staging buffer.
+// The given amount of system memory will be allocated and is accessible to both
+// the CPU and the GPU.
+//
+// Larger values here will use more memory but allow more concurrent/complex
+// command buffers. As most models that run in these environments are only a few
+// hundred dispatches per command buffer we can approximate an average
+// consumption of 500 dispatches x worst-case 256B per dispatch of parameters
+// and get 128KB.
+#define IREE_HAL_METAL_STAGING_BUFFER_DEFAULT_CAPACITY (128 * 1024)
+
+// A staging uniform buffer used for uploading parameters to the device.
+// This allows for high-frequency writes of parameters at appropriate alignment.
+//
+// Intended usage is to retain one of these per device queue and use them during
+// command buffer recording targeting that particular queue. This avoids
+// allocating a lot of small buffers. The underlying buffer has shared storage
+// mode; so it resides in system memory and is accessible to both the CPU and
+// the GPU.
+//
+// Parameters handled by this buffer include:
+// * Argument buffers for descriptor sets
+// * Source buffer for buffer update commands
+//
+// TODO(antiagainst): Use proper atomics/mutexes for concurrent command buffer
+// recording and execution.
+typedef struct iree_hal_metal_staging_buffer_t {
+  // Maximum number of bytes in the buffer.
+  uint32_t capacity;
+
+  // Device handle to the buffer.
+  id<MTLBuffer> metal_buffer;
+  // Host pointer to the buffer.
+  uint8_t* host_buffer;
+
+  // Current write offset of the device buffer.
+  uint32_t offset;
+
+  // The number of command buffers that are being recorded or executed on
```

Comment:
as you note in your TODO above this is not going to work in real programs with any level of async - we'll want to get this fixed soon as it is a correctness issue. a suballocator is the easiest solution and maybe one day I can land that and we can just swap this out with that.

Reply:
https://github.com/openxla/iree/issues/14049 to track this.